index.js

```diff
@@ -35,40 +35,40 @@ const { format } = require('node:util')
  * @returns ProcessWarning
  */
 
+/**
+ * @typedef {number} STATE_CONSTANT
+ */
+
+/**
+ * @private
+ * @typdef {object} EMISSION_STATES
+ * @property {STATE_CONSTANT} UNLIMITED_INITIAL Indicates that the warning
+ * is to be issued an unlimited number of times but has not yet been
+ * emitted.
+ * @property {STATE_CONSTANT} UNLIMITED_ONGOING Indicates that the warning
+ * is to be issued an unlimited number of times and has been emitted at least
+ * once.
+ * @property {STATE_CONSTANT} LIMITED_INITIAL Indicates that the warning
+ * is to be issued only once but has not yet been emitted.
+ * @property {STATE_CONSTANT} LIMITED_FINAL Indicates that the warning
+ * is to be issued only once and has already been emitted.
+ */
+const STATES = {
+  UNLIMITED_INITIAL: newBooleanState(0),
+  UNLIMITED_ONGOING: newBooleanState(-1),
+  LIMITED_INITIAL: newBooleanState(1),
+  LIMITED_FINAL: newBooleanState(2)
+}
```

Comment:
We might be able to make this easier to read and work with using:

```js
'use strict'

class STATE extends Boolean {
  static UNLIMITED_INITIAL = 0
  static UNLIMITED_ONGOING = -1
  static LIMITED_INITIAL = 1
  static LIMITED_FINAL = 2

  #code = 0

  constructor(value) {
    super()
    this.#code = value ?? 0
  }

  get code() {
    return this.#code
  }

  set code(val) {
    this.#code = val
  }

  isEmitted() {
    return this.valueOf()
  }

  valueOf() {
    switch (this.#code) {
      case STATE.UNLIMITED_INITIAL: return false
      case STATE.UNLIMITED_ONGOING: return true 
      case STATE.LIMITED_INITIAL: return false
      case STATE.LIMITED_FINAL: return true
    }
  }
}

const state = new STATE(STATE.UNLIMITED_INITIAL)
console.log(state)
logState(state)

state.code = STATE.UNLIMITED_ONGOING
logState(state)

function logState(s) {
  if (s.isEmitted() === false) {
    console.log('not emitted')
  } else {
    console.log('emitted')
  }
}
```

Note that I added a `.isEmitted()` because `Boolean(false) !== false`.

Reply:
Handed the work a little from @Eomm, I implemented your suggestion, can you take a look?

Although its kind of smaller codebase, its first time working on it so let me know if the changes meet the expectations 👍